ProcessLib/NeumannBC.h

```diff
@@ -0,0 +1,155 @@
+    {
+        // deep copy because the neumann bc config destroys the elements.
+        std::transform(bc->elementsBegin(), bc->elementsEnd(),
+                std::back_inserter(_elements),
+                std::mem_fn(&MeshLib::Element::clone));
+
+        std::vector<MeshLib::Node*> nodes = MeshLib::selectNodes(_elements);
+
+        _mesh_subset_all_nodes =
+            mesh_subset_all_nodes.getIntersectionByNodes(nodes);
+        _all_mesh_subsets.push_back(new MeshLib::MeshSubsets(_mesh_subset_all_nodes));
+
+        _local_to_global_index_map.reset(
+            local_to_global_index_map.deriveBoundaryConstrainedMap(
+                _all_mesh_subsets, _elements));
+    }
+
+    ~NeumannBC()
+    {
+        for (auto e : _elements)
+            delete e;
+
+        for (auto p : _local_assemblers)
+            delete p;
+    }
+
+    template <typename GlobalSetup>
+    void
+    initialize(GlobalSetup const& global_setup,
+        typename GlobalSetup::MatrixType& A,
+        typename GlobalSetup::VectorType& rhs)
+    {
+        // Shape matrices initializer
+        using LocalDataInitializer = AssemblerLib::LocalDataInitializer<
+            LocalNeumannBcAsmDataInterface,
+            LocalNeumannBcAsmData,
+            typename GlobalSetup::MatrixType,
+            typename GlobalSetup::VectorType>;
+
+        LocalDataInitializer initializer;
+
+        using LocalAssemblerBuilder =
+            AssemblerLib::LocalAssemblerBuilder<
+                MeshLib::Element,
+                LocalDataInitializer>;
+
+        // Populate the vector of local assemblers.
+        _local_assemblers.resize(_elements.size());
+        LocalAssemblerBuilder local_asm_builder(
+            initializer, *_local_to_global_index_map);
+
+        auto elementValueLookup = [this](MeshLib::Element const& e)
+        {
+            return _function() * _scaling_value;
+        };
+
+        DBUG("Calling local Neumann assembler builder for neumann boundary elements.");
+        global_setup.execute(
+                local_asm_builder,
+                _elements,
+                _local_assemblers,
+                elementValueLookup,
+                _integration_order);
+
+        DBUG("Create global assembler.");
+        _global_assembler.reset(
+            new GlobalAssembler(A, rhs, *_local_to_global_index_map));
+    }
+
+    template <typename GlobalSetup>
+    void
+    integrate(GlobalSetup const& global_setup)
+    {
+        global_setup.execute(*_global_assembler, _local_assemblers);
+    }
+
+
+private:
+    double const _scaling_value;
+
+    MathLib::ConstantFunction<double> const _function;
+
+    std::vector<MeshLib::Element const*> _elements;
+    MeshLib::MeshSubset const* _mesh_subset_all_nodes = nullptr;
+    std::vector<MeshLib::MeshSubsets*> _all_mesh_subsets;
+
+    std::unique_ptr<AssemblerLib::LocalToGlobalIndexMap> _local_to_global_index_map;
+
+    unsigned const _integration_order;
+
+    using GlobalAssembler =
```

Comment:
It looks like that including MatrixType here is just for a unqiue interface. Am I right?